File: day1/day1a.rb
# frozen_string_literal: true

puts ARGF
  .read.scan(/\d+/)
  .map(&:to_i)
  .each_slice(2)
  .to_a
  .transpose
  .map(&:sort)
  .transpose
  .map { (_1 - _2).abs }
  .sum

File: day1/day1a.clj
#!/usr/bin/env clojure -M

(->> (slurp *in*)                              ; Read all input from stdin
       (re-seq #"\d+")                            ; Extract all digit sequences
       (map read-string)                          ; Convert strings to integers
       (partition 2)                              ; Group numbers into pairs
       (apply map vector)                         ; Transpose the list of pairs
       (map sort)                                 ; Sort each column
       (apply map vector)                         ; Transpose back to pairs
       (map (fn [[x y]] (abs (- x y))))           ; Calculate absolute differences
       (reduce +)                                 ; Sum all differences
       (println))                                 ; Print the result

